krun/scheduler.py

```diff
@@ -279,160 +314,178 @@ def _make_pre_post_cmd_env(self):
-                # reboot mode.
-                eta_info += STARTUP_WAIT_SECONDS
-            self.add_eta_info(job.key, eta_info)
-
-            info("%d executions left in scheduler queue" % (jobs_left - 1))
-
-            # We dump the json after each experiment so we can monitor the
-            # json file mid-run. It is overwritten each time.
-            self.results.write_to_file()
-
-            self.jobs_done += 1
-
-            # Run the user's post-process-execution commands with updated
-            # ETA estimates. Important that this happens *after* dumping
-            # results, as the user is likely copying intermediate results to
-            # another host.
-            util.run_shell_cmd_list(
-                self.config.POST_EXECUTION_CMDS,
-                extra_env=self._make_pre_post_cmd_env()
-            )
+        # No executions, or all skipped
+        if self.manifest.num_execs_left == 0:
+            info("Empty schedule!")
+            return
 
-            tfmt = self.get_overall_time_estimate_formatter()
+        self.platform.wait_for_temperature_sensors()
 
-            if self.eta_avail == self.jobs_done:
-                # We just found out roughly how long the session has left, mail out.
-                msg = "ETA for current session now known: %s" % tfmt.finish_str
-                util.log_and_mail(self.mailer, debug,
-                             "ETA for Current Session Available",
-                             msg, bypass_limiter=True)
-
-            info("{:<25s}: {} ({} from now)".format(
-                "Estimated completion", tfmt.finish_str, tfmt.delta_str))
-
-            try:
-                job = self.next_job(peek=True)
-            except ScheduleEmpty:
-                pass  # no next job
-            else:
-                # print info about the next job
-                info("Next execution is '%s(%d)' (%s variant) under '%s'" %
-                     (job.benchmark, job.parameter, job.variant, job.vm_name))
-
-                tfmt = self.get_exec_estimate_time_formatter(job.key)
-                info("{:<35s}: {} ({} from now)".format(
-                    "Estimated completion (next execution)",
-                    tfmt.finish_str,
-                    tfmt.delta_str))
-
-            if (self.eta_avail is not None) and (self.jobs_done < self.eta_avail):
-                info("Executions until ETA known: %s" % self.jobs_until_eta_known())
-
-            if self.platform.check_dmesg_for_changes():
-                self.results.error_flag = True
-
-            if self.reboot and len(self) > 0:
-                info("Reboot in preparation for next execution")
-                self._reboot()
+        if self.on_first_invokation:
+            info("Reboot prior to first execution")
+            self._reboot()
 
-        self.platform.save_power()
+        # If we get here, this is a real run, with a benchmark about to
```

Comment:
I'm really happy with the simplification of `run()`, it's much easier to understand now.